src/ngx_http_lua_output.c

```diff
@@ -802,4 +802,99 @@ ngx_http_lua_flush_cleanup(void *data)
     ctx->flushing_coros--;
 }
 
+#ifndef NGX_LUA_NO_FFI_API
+//Pre-Condition: offset + len <= size of buf
+int
+ngx_http_lua_ffi_write(ngx_http_request_t *r, const char *buf, size_t offset,
+                       size_t len, char **err)
+{
+    ngx_http_lua_ctx_t          *ctx;
+    ngx_buf_t                   *b;
+    ngx_chain_t                 *cl;
+    ngx_int_t                    rc;
+    const char                  *str_sub;
+
+    if (r == NULL) {
+        *err = "no request object found";
+        return NGX_ERROR;
+    }
+
+    if (buf == NULL) {
+        *err = "no buffer found";
+        return NGX_ERROR;
+    }
+
+    if (len == 0) {
+        *err = "bad len from offset";
+        return NGX_ERROR;
+    }
+
+    str_sub = buf + offset;
+
+    if (r->connection->fd == (ngx_socket_t) -1) {
+        *err = "bad connection context";
+        return NGX_HTTP_LUA_FFI_BAD_CONTEXT;
+    }
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+
+    if (ctx == NULL) {
+        *err = "no request context found";
+        return NGX_HTTP_LUA_FFI_NO_REQ_CTX;
+    }
+
+    if (!(ctx->context & (NGX_HTTP_LUA_CONTEXT_REWRITE
+                          | NGX_HTTP_LUA_CONTEXT_ACCESS
+                          | NGX_HTTP_LUA_CONTEXT_CONTENT))) {
+        *err = "API disabled in the current context";
+        return NGX_HTTP_LUA_FFI_BAD_CONTEXT;
+    }
+
+    if (ctx->acquired_raw_req_socket) {
+        *err = "raw request socket acquired";
+        return NGX_ERROR;
+    }
+
+    if (r->header_only) {
+        ctx->eof = 1;
+        *err = "header only";
+        return NGX_ERROR;
+    }
+
+    if (ctx->eof) {
+        *err = "seen eof";
+        return NGX_ERROR;
+    }
+
+    ctx->seen_body_data = 1;
+
+    cl = ngx_http_lua_chain_get_free_buf(r->connection->log, r->pool,
+                                         &ctx->free_bufs, len);
+
+    if (cl == NULL) {
+        *err = "no memory";
+        return NGX_ERROR;
+    }
+
+    b = cl->buf;
+    b->last = ngx_copy(b->last, (u_char *) str_sub, len);
+
+    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+            "lua ffi_write response");
+
+    rc = ngx_http_lua_send_chain_link(r, ctx, cl);
+
+    if (rc == NGX_ERROR || rc >= NGX_HTTP_SPECIAL_RESPONSE) {
+        *err = "output filter error";
+        return NGX_ERROR;
+    }
+
+    dd("downstream write: %d, buf len: %d", (int) rc,
+            (int) (b->last - b->pos));
```

Comment:
There are some remaining coding style issues like this, which can be detected by the [ngx-releng](https://github.com/openresty/openresty-devel-utils/blob/master/ngx-releng) tool. Please use the tool to make this C file clean of such minor issues. Thanks!